Added Longest Prefix Match to FMDIndex

cramertj opened the pull request:
Tests aren't super comprehensive, but I added longest prefix matching to the FMDIndex, did a small bit of reformatting, and factored out common code.

johanneskoester commented on src/data_structures/fmindex.rs:
```diff
@@ -233,7 +276,11 @@ impl FMDIndex {
         let mut j = pattern.len() as isize;
 
         for k in (-1..i as isize).rev() {
-            let a = if k == -1 { b'$' } else { pattern[k as usize] };
```
Do you think this is bad style? If so, is there a style guide suggesting this?

johanneskoester commented on src/data_structures/fmindex.rs:
```diff
@@ -168,18 +183,46 @@ impl FMDIndex {
 
+    /// Find longest non-encompassing prefix match
+    pub fn longest_prefix_match(&self, pattern: &[u8], min_overlap: usize) -> Option<BiInterval> {
+        let exact_match_interval = self.fmindex.backward_search(pattern.iter());
```
My question, can't we just do a backward search from the reverse pattern, stopping at an empty interval or before reading the first character? Then, the longest prefix can be can be easily returned, can't it?

cramertj replied:
Yes, that seems much more reasonable. Thanks for the suggestion. I'll make the change.

cramertj replied:
Ah--- wouldn't that include exact matches, though?

johanneskoester replied:
Not if you stop before reading the first character of the reverse pattern (which is the last character of the original pattern).

cramertj replied:
So in that case, how do I make sure that as I'm continuing, I still have valid prefixes (i.e. they end in $ before hitting my string)? To clarify, we're looking for the string ACGT$, but not ACGTA..., when our pattern is CGTG$.

cramertj replied:
Sorry if my inexperience is showing on this issue-- I admit that I have yet to fully comprehend some of the inner workings of the FMIndex functions. I'm confident my solution works and runs in O(m) time (m -> length of pattern), but I'm willing to accept that the backtracking through the results at the end looking for the ones that can extend to contain the sentinel may not be the most efficient method :).

johanneskoester replied:
I am not sure that I entirely agree with your use of the terminology here. What you are talking about, aren't that rather suffixes which are prefixes of the pattern? Thinking about it, that makes also much more sense in the string graph context you mentioned before...

johanneskoester replied:
Maybe algorithm 3 in section 2.5 of [this book](ftp://ftp.sanger.ac.uk/pub/resources/theses/js18/chapter2.pdf) is what you think about?

johanneskoester replied:
Then, there is also this paper:
http://bioinformatics.oxfordjournals.org/content/26/12/i367.full

cramertj replied:
What I'm doing is very similar to the algorithm given on page 33 of the Sanger paper.

cramertj replied:
If I change the name of the function to something more transparent (perhaps longest_prefix_suffix_match) and add some docs, are we good to go?

johanneskoester replied:
So, one of the design principles of Rust-Bio is to use the state of the
art algorithm if it exists. So, since there are publications available,
I would like the implementation to follow them. If your implementation
is pretty similar already, it should be easy to adapt. I'm really sorry
for bothering you with this, but I just want to codebase to follow this
rule whenever possible. This is also important to give credit to the
right people in the docs, instead of slightly reinventing the wheel.
Also, note that both pubs are about the FMIndex, not the FMDIndex. I am
not sure about the implications of doing this on the FMDIndex, nor am I
sure if this is wanted in your case (it will also consider reverse
complements).
On Tue, 2016-03-01 at 09:31 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> What I'm doing is very similar to the algorithm given on page 33 of
> the Sanger paper.
> —
> Reply to this email directly or view it on GitHub.

cramertj replied:
Okay. I've tested it, and it does what I expected (it's fine that it considers reverse complements-- it should, since reads can occur from either direction during sequencing). I've committed a new version with the name change, and also changed it to return the length of the overlap. I can take a look later and try and match it more precisely to the Sanger pseudocode, but it probably won't be until later this week or early next week. If you have time yourself, feel free to make whatever changes you see fit.

johanneskoester replied:
Thanks, I'm glad you understand my point. Take your time, I can wait.
On Tue, 2016-03-01 at 13:01 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> Okay. I've tested it, and it does what I expected (it's fine that it
> considers reverse complements-- it should, since reads can occur from
> either direction during sequencing). I've committed a new version
> with the name change, and also changed it to return the length of the
> overlap. I can take a look later and try and match it more precisely
> to the Sanger pseudocode, but it probably won't be until later this
> week or early next week. If you have time yourself, feel free to make
> whatever changes you see fit.
> —
> Reply to this email directly or view it on GitHub.

cramertj replied:
Additionally, what was your design intention with splitting up FMIndex and FMDIndex the way you did? I agree that the function I provided still makes sense in the context of an FMIndex, but if you create an FMDIndex (i.e. an FMIndex with DNA of the pattern F$R$, in my mind), you should still be able to perform all of the normal FMIndex functionality, right?

cramertj replied:
It seems strange that the behavior of an FMIndex is exclusively backward_search, and smems for FMDIndex. There are many more algorithms that exist for these/this data structure/s, so I'm not sure why the underlying functionality (occ, less, init, forward/backward extend) is encapsulated without also providing these algorithms. Essentially, it makes it so that client developers (like me :) ) have to change the underlying library code to get the algorithm they want. Perhaps the eventual goal is to have inbuilt support for all common FMIndex/FMDIndex algorithms, so that clients won't have to implement them?

Anyway. This is definitely getting out of scope for this particular PR. Thanks for your help-- I'll get something together for a more by-the-book algorithm soon.

johanneskoester replied:
Hi Taylor,
those are very valid questions.
The rationale was that an FMIndex is a know DS, and therefore should be
a single object that owns all its parts (bwt, occ, less). This way, it
can be easily passed to other structs etc. Of course I did not have the
time so far to implement all possible algorithms on top of the FMIndex.
When sb wants to do that, he has still all possibilities on the client
side, since occ, less and bwt of the FMIndex are not hidden away but
rather available via getters (however I just noticed that I
accidentally missed the pub before occ and less!!!).
Coming to FMDIndex, it is really an extension of FMIndex that only
works on a special case, namely a text with forward and reverse
complement concatenated. Further, it is only defined on the DNA
alphabet, whereas the FMIndex is defined for arbitrary alphabets.
Hence, smems cannot be a member of FMIndex. Again, with FMDIndex, the
underlying FMIndex is accessiable via a getter, so extension is
possible without modifying the struct itself.
But of course, I would like all standard/textbook applications of FM
and FMDIndex to be in Rust-Bio. I just don't have the resources to do
that myself unless I need a particular algorithm for my own work.
However, of course I would be interested in alternative API solutions
if you have one. For example, one could think of a common trait for the
two structs. This would allow the user to conveniently add other
methods via custom traits, similar to the itertools package.
Best,
Johannes
On Tue, 2016-03-01 at 13:14 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> Additionally, what was your design intention with splitting up
> FMIndex and FMDIndex the way you did? I agree that the function I
> provided still makes sense in the context of an FMIndex, but if you
> create an FMDIndex (i.e. an FMIndex with DNA of the pattern F$R$, in
> my mind), you should still be able to perform all of the normal
> FMIndex functionality, right?
> —
> Reply to this email directly or view it on GitHub.

cramertj replied:
Ah, so those were supposed to be pub... That makes much more sense now. :)

cramertj replied:
Would it maybe make sense to break out FMIndex into a trait and say that both the FMIndex and FMDIndex structs _are_ FMIndexes?

johanneskoester replied:
I have fixed them in the master branch.
On Wed, 2016-03-02 at 12:28 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> Ah, so those were supposed to be pub... That makes much more sense
> now. :)
> —
> Reply to this email directly or view it on GitHub.

johanneskoester replied:
Maybe, sounds reasonable at first sight. I have to further think about
it.
On Wed, 2016-03-02 at 12:30 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> Would it maybe make sense to break out FMIndex into a trait and say
> that both the FMIndex and FMDIndex structs are FMIndexes?
> —
> Reply to this email directly or view it on GitHub.

johanneskoester replied:
Yes, that's a nice idea! I like it.
On Wed, 2016-03-02 at 12:30 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> Would it maybe make sense to break out FMIndex into a trait and say
> that both the FMIndex and FMDIndex structs are FMIndexes?
> —
> Reply to this email directly or view it on GitHub.

johanneskoester commented on src/data_structures/fmindex.rs:
```diff
@@ -81,15 +85,25 @@ impl FMIndex {
         let (mut l, mut r) = (0, self.bwt.len() - 1);
         for &a in pattern.rev() {
             let less = self.less(a);
-            l = less + if l > 0 { self.occ(l - 1, a) } else { 0 };
```
Is this bad style?

johanneskoester commented on src/data_structures/fmindex.rs:
```diff
@@ -81,15 +85,25 @@ impl FMIndex {
             r = less + self.occ(r, a) - 1;
         }
 
-        Interval { lower: l, upper: r + 1 }
```
See above.

cramertj commented:
The style changes are mostly a result of https://github.com/rust-lang-nursery/rustfmt. Feel free to tell me to revert any/all.

johanneskoester commented:
I wasn't aware of that, cool. In that case, let's follow the standard. I will run it on the whole rust-bio soon.

cramertj commented:
I've added a test case for suffix_array that's currently failing. I fixed a bug in suffix array's "transform_text" fn that I think was causing the problem, but I'm working on confirming before I push.

cramertj commented:
Fixed-- there were off-by-one issues in suffix_array.rs

johanneskoester commented:
Can you extract the suffix array fix into a separate PR? Thanks for noticing and the fix!

cramertj commented:
Done!

johanneskoester commented:
Thanks!

cramertj commented:
Closing for re-evaluation
